openapi.yaml

```
@@ -0,0 +1,2411 @@
+
+  /dags:
+    get:
+      summary: Get all DAGs
+      operationId: getDags
+      tags: [DAG]
+      parameters:
+        - $ref: '#/components/parameters/PageLimit'
+        - $ref: '#/components/parameters/PageOffset'
+      responses:
+        '200':
+          description: List of DAGs.
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/DAGCollection'
+                  - $ref: '#/components/schemas/CollectionInfo'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+
+  /dags/{dag_id}:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+
+    get:
+      summary: Get basic information about a DAG
+      description: Presents only information available in database (DAGModel).
+      operationId: getDag
+      tags: [DAG]
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/DAG'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+    patch:
+      summary: Update a DAG
+      operationId: updateDag
+      tags: [DAG]
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/DAG'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+  /dags/{dag_id}/clearTaskInstanaces:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+
+    post:
+      summary: Clears a set of task instances associated with the DAAG for a specified date range.
+      operationId: clearTaskInstaance
+      tags: [DAG]
+      requestBody:
+        description: Parameters of action
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ClearTaskInstance'
+
+      responses:
+        '200':
+          description: A list of cleared tasks
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TaskInstanceCollection'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+  /dags/{dag_id}/dagRuns:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+
+    get:
+      summary: Get all DAG Runs
+      description: >
+        This endpoint support reading resources across multiple DAGs by specifying a "~" as a dag_id.
```

My complaint against `-` was based on a mis-understanding, where I thought we were doing `dag_id1-dag_id2`.

This may actually be a bad choice (sorry!) as `~` isn't URL safe and would be `%7E`.  Obviously that's workable, but, how about:

```suggestion
        This endpoint allows specifying `-` as the dag_id to retrieve DAG Runs for all DAGs
```

and similarly for other examples. I think I was confused by the word multiple (compared to "all")

Technically this forbids the use of `-` as a DAG id. I think I'm okay with disallowing that one specific case (cos really, who would do that.)

Would you like me to bring this up on the mailing list thread for other peoples' views?

---

We discussed this topic on the dev list. https://lists.apache.org/thread.html/rc38353666b5d6bf3c4dce3b0af04a82e57591fefc7da171f4535947f%40%3Cdev.airflow.apache.org%3E
Should I start a new thread?

---

I am also good with '-' (dash) if nobody complains about it's ambiguity (it can be valid dag_id).

I proposed '~' instead of '*' as '*' is one of the reserved characters in URL and different clients might treat it differently depending on if it is encoded or not (it's already happening in different languages). Its behaviour depends on the URL scheme/client. 

According to RFC3986 this is expected that different URL schemes treat encoded/unencoded '*' differently (and few other reserved characters): 
reserved    = gen-delims / sub-delims
      gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"
      sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="

From: https://tools.ietf.org/html/rfc3986#section-2.2 

I understood that the problem with '-' was that you can potentially have dag name '-' (I believe it's a valid dag name) so it would introduce ambiguity. But it is really a nuance and we can safely assume that no one sane would use '-' as DAG name. 

@ashb:  actually '~'  is an unreserved character (together with the dash, underscore, dot, alphas, and digits so no matter if it's urlencoded or not, all clients and servers should treat it the same way (so it never has to be urlencoded): 

      unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"

From: https://tools.ietf.org/html/rfc3986#section-2.3

---

> We discussed this topic on the dev list. https://lists.apache.org/thread.html/rc38353666b5d6bf3c4dce3b0af04a82e57591fefc7da171f4535947f%40%3Cdev.airflow.apache.org%3E
> I will start a new thread.

Sure, but I think that was driven by my misunderstanding.

Ah, if it's unreserved by that, and shell won't replace it in `/dag/~/dag_runs/` then I'm happy to leave it as it is. (I don't care if something escapes it unnecessarily, I was thinking about `curl` use cases from command line, and having it "just work" which is sounds like it will.

This can stay as it is then. :+1:

---

We are all on one page. I wonder if we need another mailing list discussion. @ashb WDYT?

---

Not on this issue, I think we're good here.

I'd like to check over the end of the thread first before we merge this please.

Giving myself a deadline: if I haven't otherwise replied by end of tomorrow, consider this my :+1: